update and cleanup contributing guidelines (#73)
ml5js/ml5-next-gen: CONTRIBUTING.md

# How to Contribute

_last updated: 3, July 2024_

> Welcome to the ml5 project! Developing ml5 is not just about developing machine learning software, it is about making machine learning approachable for a broad audience of artists, creative coders, and students. The library provides access to machine learning algorithms and models in the browser, building on top of TensorFlow.js with no other external dependencies. The library is supported by code examples, tutorials, and sample datasets with an emphasis on ethical computing. Bias in data, stereotypical harms, and responsible crowdsourcing are part of the documentation around data collection and usage. We're building friendly machine learning for the web - we're glad you're here!

Reviewer: ```suggestion
Welcome to the ml5.js project! Developing ml5.js is not just about developing machine learning software, it is about making machine learning approachable for a broad audience of artists, creative coders, and students. The library provides access to machine learning algorithms and models in the browser, building on top of TensorFlow.js with no other external dependencies. The library is supported by code examples, tutorials, and sample datasets with an emphasis on ethical computing. Bias in data, stereotypical harms, and responsible crowdsourcing are part of the documentation around data collection and usage. We're building friendly machine learning for the web - we're glad you're here!
```

tiny thing but the preferred style is to include the `js`

## Table of Content

- [The ml5.js Ecosystem](#the-ml5js-ecosystem)
- [Contributing Workflow](#contributing-workflow)
- [Setup Development Environment](#setup-development-environment)
- [Code Formatting](#code-formatting)
- [Building the Library](#building-the-library)
- [Making Releases](#making-releases)
- [Unit Tests](#unit-tests)
- [Update p5 Web Editor Sketches](#update-p5-web-editor-sketches)
- [All Contributors](#all-contributors)
- [API Implementation Guideline](#api-implementation-guideline)
- [Utils](#utils)

## The ml5.js Ecosystem

> ml5.js is comprised a number of sister repositories which you can find at the ml5 github organization - [github.com/ml5js](https://github.com/ml5js). As a contributor of ml5 you should be aware of the other parallel repositories of the ml5 project.

Reviewer: ```suggestion
ml5.js is comprised a number of related repositories which you can find at the ml5.js github organization - [github.com/ml5js](https://github.com/ml5js). As a contributor of ml5.js you should be aware of the other parallel repositories of the ml5 project.
```

## Contributing Workflow

**Preamble**: If you're interested in contributing to the ml5 project, just know you can always open an issue to ask questions or flag things that may seem confusing, unclear or intimidating. Our goal is to make ml5 as open and supportive as possible for those who want to be involved. Ok, now that's out of the way, here's how a general workflow for what contributions might look like to ml5.

1. Read the CONTRIBUTING.md docs. ❤️
2. Take a peek at the [issues](https://github.com/ml5js/ml5-next-gen/issues) and identify something you'd like to address **or** file an issue about a bug you discovered. The issue could be addressing a bug, adding a new feature, updating an existing feature, or anything else. 🚩
3. Make a comment on the existing issue **or** indicate on your new issue that you're curious to do your best to solve it. 🔬
4. Make a `forked` copy of the ml5-library and create it branch with a meaningful name such as `fix-detection-results`. 🍴

> 5. Jam on some coding sessions, commit your changes with meaningful commit messages, and push your changes to your branch. 💻

Reviewer: ```suggestion
5. Jam on some coding sessions, commit your changes with descriptive commit messages, and push your changes to your branch. 💻
```

6. When ready, [make a pull request to the `main` branch of ml5-next-gen](https://github.com/ml5js/ml5-next-gen/compare). 📄

> 7. The ml5 dev team will review your changes and quite likely correspond with you on your changes. When all looks good, your changes will be merged in and released with the next public update to the library. 🎉

Reviewer: ```suggestion
7. The ml5.js dev team will review your changes and reply with feedback or questions. When all looks good, your changes will be merged in and released with the next public update to the library. 🎉
```

This will create a production version of the library in `/dist` directory.

## API Implementation Guideline

> This guideline provides a high-level concept of what the ml5 library's API should look like, serving as a reference to keep the ml5 interface consistent and accessible to the users.

Reviewer: ```suggestion
This guideline provides a high-level concept of what the ml5 library's API should look like, serving as a reference to keep the ml5 interface consistent and friendly.

```

webpack 5.76.1 compiled successfully in 8360 ms
```

> A local server have been started, hosting a built version of the ml5 library at http://localhost:8080/dist/ml5.js. While the server is running, Webpack will automatically rebuild the library if you change and save any file in the `/src` folder.

Reviewer: ```suggestion
A local server will start, hosting a built version of the ml5.js library at http://localhost:8080/dist/ml5.js. While the server is running, Webpack will automatically rebuild the library if you change and save any file in the `/src` folder.
```

npx prettier --write path/to/file
```

For more options with the command line, please refer to the [Prettier Documentation](https://prettier.io/docs/en/cli.html)

## Building the Library

> To build the ml5 library for production and publishing, run the following commands:

Reviewer: ```suggestion
To build the ml5.js library for production and publishing, run the following commands:
```

This will create a production version of the library in `/dist` directory.

> ## Unit Tests

Reviewer: Maybe we should write something here for how we only have placeholder unit tests but this is something we are investigating and hope to add in the future. We could encourage anyone with experience or knowledge about unit testing to open an issue for discussion. cc @sharellb

Author: I think this is a good idea!

Make a pull request to merge the new branch into main. When the branch gets merged, the new contributor will show up in README.md!

## API Implementation Guideline

> This guideline provides a high-level concept of what the ml5 library's API should look like, serving as a reference to keep the ml5 interface consistent and friendly.

Reviewer: ```suggestion
This guideline provides a high-level concept of what the ml5.js library's API should look like, serving as a reference to keep the ml5.js interface consistent and friendly.
```

> (Temporary note: This section does not fully reflect the current interface of the library, but rather proposes an ideal interface for future ml5.)

Reviewer: Is this still true?

Author: Ah, I will take a look.

(Temporary note 2: This section is work in progress.)

### General Guidelines

> 1. All ml5 functions called by the user should be defined in camel case.

Reviewer: ```suggestion
1. All ml5.hs functions called by the user are defined with "camelCase".
```

> 2. The user accesses an instance of the ml5 model by calling the factory function `ml5.<modelName>`. For example:

Reviewer: ```suggestion
2.  All ml5.js models are instantiated with a factory function with the style `ml5.<modelName>`. For example:
```

```javascript
let bodyPose = ml5.bodyPose("BlazePose");
let neuralNetwork = ml5.neuralNetwork();
```

> 4. When a string parameter is passed into any ml5 function as a configuration setting, it should be matched case insensitively For example, the following calls produce the same result:

Reviewer: ```suggestion
4. When a string parameter is passed into any ml5.js function as a configuration setting, it is matched case _insensitively_ . For example, the following calls produce the same result:
```

   ```javascript
   let bodyPose = ml5.bodyPose("BlazePose", { modelType: "Full" });
   ```

   ```javascript
   let bodyPose = ml5.bodyPose("blazepose", { modelType: "full" });
   ```

> 5. When possible, a parameter should be optional and have default values in lieu of user definition.

Reviewer: ```suggestion
5. When possible, all parameters should be optional and have default values in lieu of user definition.
```

> 6. ml5 should throw a friendly warning when the user passes in an invalid parameter, and proceed with the default value when possible. For example:

Reviewer: ```suggestion
6. ml5.js should throw a friendly warning when the user passes in an invalid parameter, and proceed with the default value when possible. For example:
```

   ```javascript
   let bodyPose = ml5.bodyPose({ modelType: "foo" });
   // Console:
   // Warning: bodyPose does not have modelType "foo", using modelType "full" instead.
   ```
7. An ml5 function should be able to accept parameters in any order. For example, the following calls produce the same result:

   ```javascript
   let bodyPose = ml5.bodyPose("blazepose", { modelType: "full" });
   ```

   ```javascript
   let bodyPose = ml5.bodyPose({ modelType: "full" }, "blazepose");
   ```

> 8. Asynchronous ml5 functions should support the callback interface, and support the promise/async await interface as well when possible.

Reviewer: ```suggestion
8. Asynchronous ml5 functions should support the callback interface, and support the promise/async await interface as well when possible.
```

Are we supporting promise/async/await with the models? maybe this should say something like. "The ml5.js models do not currently support promises (async/await), however, we are considering adding this in the future and plan to follow the p5.js's lead."

Author: Most of the current code does support the promise interface when possible. Perhaps we can say something along the lines of "Although some existing code has both callback and promise interface, the ml5.js models do not currently support promises (async/await) officially, however, we are considering adding this in the future and plan to follow the p5.js's lead."

9. ml5's should call the callback functions with the pattern `callback(result, error)`.

> 10. The factory functions `ml5.<modelName>` should support p5's preload feature as well as the callback interface. Thus, the factory functions should synchronously/immediately return an instance of the ml5 model and load the underlying model asynchronously before calling the callback/releasing p5's preload hook.

Reviewer: ```suggestion
10. The factory functions `ml5.<modelName>` should support p5.js's `preload()` as well as the callback interface. When called in `preload()`, the factory functions should synchronously return an instance of the ml5.js model. Otherwise, the model is loaded asynchronously before calling the callback.
```

I'm not sure if my revisions here are accurate, please check!

Author: Yes, I think the factory function returns an unready instance of the model and it will either "let `preload` proceed" or call the callback when the model instance becomes ready. The same instance is also passed into the callback. 

I will double-check the code to make sure!
